OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs

```diff
@@ -68,57 +84,113 @@ public void Tick(Squad owner)
+		int makeWay = MakeWayTicks;
+		WPos lastPos = WPos.Zero;
+
 		public void Activate(Squad owner) { }
 
 		public void Tick(Squad owner)
 		{
+			// Basic check
 			if (!owner.IsValid)
 				return;
 
 			if (!owner.IsTargetValid)
 			{
-				var closestEnemy = FindClosestEnemy(owner);
-				if (closestEnemy != null)
-					owner.TargetActor = closestEnemy;
+				var targetActor = GetRandomPreferredTarget(owner);
+				if (targetActor != null)
+					owner.TargetActor = targetActor;
 				else
 				{
 					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsFleeState(), true);
 					return;
 				}
 			}
 
-			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
+			// Initialize leader. Optimaze pathfinding by using leader.
+			var leader = owner.Units.FirstOrDefault();
 			if (leader == null)
 				return;
 
-			var ownUnits = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.Units.Count) / 3)
-				.Where(a => a.Owner == owner.Units.First().Owner && owner.Units.Contains(a)).ToHashSet();
+			// 1. Scanning if there is enemy around and switch to GroundUnitsAttackState if any.
+			var attackScanRadius = WDist.FromCells(owner.SquadManager.Info.AttackScanRadius);
 
-			if (ownUnits.Count < owner.Units.Count)
+			var enemyActor = ScanThreat(owner, leader, attackScanRadius);
+			if (enemyActor != null)
 			{
-				// Since units have different movement speeds, they get separated while approaching the target.
-				// Let them regroup into tighter formation.
-				owner.Bot.QueueOrder(new Order("Stop", leader, false));
-				foreach (var unit in owner.Units.Where(a => !ownUnits.Contains(a)))
-					owner.Bot.QueueOrder(new Order("AttackMove", unit, Target.FromCell(owner.World, leader.Location), false));
+				owner.TargetActor = enemyActor;
+				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+				return;
 			}
-			else
+
+			// 2. Make sure the guide unit has not been blocked by the rest of the squad
+			if (attemptsToAdvance <= 0)
 			{
-				var enemies = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.SquadManager.Info.AttackScanRadius))
-					.Where(owner.SquadManager.IsPreferredEnemyUnit);
-				var target = enemies.ClosestTo(leader.CenterPosition);
-				if (target != null)
+				if (makeWay > 0)
 				{
-					owner.TargetActor = target;
-					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+					owner.Bot.QueueOrder(new Order("AttackMove", leader, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+
+					foreach (var a in owner.Units)
+					{
+						if (a != leader)
+							owner.Bot.QueueOrder(new Order("Scatter", a, false));
+					}
+
+					makeWay--;
 				}
 				else
-					foreach (var a in owner.Units)
-						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				{
+					// When going through is over, restore the check
```

Comment:
I don't understand what this comment is trying to say. I expect that fixing `attemptsToAdvance` to count up should make this code clear enough to be self-explanatory, and this can probably just be removed.

Reply:
removed.